timeboost-sequencer/src/decrypt.rs

```diff
@@ -570,115 +639,173 @@ impl Worker {
+            self.state = WorkerState::ResharingPending(dec_sk);
+            info!(node = %self.label, committee_id = %committee.id(), "dkg finished (catchup successful)");
+        }
+
+        Ok(())
+    }
 
-                self.enc_key.set(dec_sk.pubkey().clone());
-                self.dkg_state = DkgState::Completed(dec_sk);
-                info!(committee_id = %committee.id(), node = %self.label, "DKG finished (node successfully recovered)");
+    /// A resharing message has been received.
+    async fn on_resharing_msg(&mut self, src: PublicKey, msg: ReshareMessage) -> Result<()> {
+        trace!(node = %self.label, from=%src, %msg.committee_id, "received resharing message");
+        if msg.committee_id != self.current {
+            info!(node = %self.label, current = %self.current, %msg.committee_id, "not current committee");
+            return Ok(());
+        }
+
+        let subsets = match &mut self.state {
+            WorkerState::AwaitingHandover(subsets) => subsets,
+            _ => {
+                trace!(node = %self.label, current = %self.current, %msg.committee_id, "not awaiting handover");
+                return Ok(());
             }
+        };
+
+        let guard = self.key_stores.read();
+        if guard.len() < 2 {
+            warn!(node = %self.label, current = %self.current, %msg.committee_id, "no previous key store");
+            return Ok(());
+        }
+        let (prev, current) = (guard.last(), guard.first());
+
+        subsets.insert(src, msg.subset.to_owned());
+
+        let threshold: usize = prev.committee().one_honest_threshold().into();
+
+        let mut counts = HashMap::new();
+        for subset in subsets.values() {
+            *counts.entry(subset).or_insert(0) += 1;
+        }
+
+        if let Some((&subset, _)) = counts.iter().find(|(_, count)| **count >= threshold) {
+            let committee = current.committee();
+            let vess = Vess::new_fast();
+            let dealings: Vec<_> = subset
+                .bundles()
+                .iter()
+                .enumerate()
+                .map(|(i, b)| {
+                    let node_idx = b.origin().0.into();
+                    let pub_share = subset.combkey().get_pub_share(node_idx).ok_or_else(|| {
+                        DecrypterError::Internal(format!(
+                            "missing pub share for node idx: {node_idx}",
+                        ))
+                    })?;
+                    vess.decrypt_reshare(committee, &self.dkg_sk, b.vess_ct(), DKG_AAD, *pub_share)
+                        .map(|s| (i, s, b.comm().clone()))
+                        .map_err(|e| DecrypterError::Dkg(format!("unable to decrypt reshare: {e}")))
+                })
+                .collect::<Result<Vec<_>>>()?;
+
+            let next_dec_key = DecryptionKey::from_resharing(
+                prev.committee(),
+                committee,
+                self.dkg_sk.node_idx(),
+                dealings.into_iter(),
```

Comment:
do we really need to collect this into `dealing`? can we use the `ResultIter` trick to save this collection? following the pattern in `on_dkg_response_msg()`

Reply:
ignore this comment, I will try to address it in future PR: https://github.com/EspressoSystems/timeboost/pull/441#discussion_r2275592440